cores/nRF5/Tone.cpp

```diff
@@ -34,161 +34,397 @@ Version Modified By Date     Comments
+            this->task_to_start = NRF_PWM_TASK_SEQSTART0;
+        }
+        this->loop_count   = full_loops + 1;
+        this->seq0_refresh = seq0_count - 1;
+        this->seq1_refresh = (total_refresh_count - seq0_count) - 1;
+        this->shorts        = NRF_PWM_SHORT_LOOPSDONE_STOP_MASK;
+    }
+    this->is_initialized = true;
+    return true;
 }
-#endif
-
-void PWM2_IRQHandler(void){
-	nrf_pwm_event_clear(NRF_PWM2, NRF_PWM_EVENT_PWMPERIODEND);
-	if(!no_stop){
-		count_duration--;
-		if(count_duration == 0)
-			noTone(pin_sound);
-		else
-			nrf_pwm_task_trigger(NRF_PWM2, NRF_PWM_TASK_SEQSTART0);	
-	}
-	else
-		nrf_pwm_task_trigger(NRF_PWM2, NRF_PWM_TASK_SEQSTART0);
+bool TONE_PWM_CONFIG::ApplyConfiguration(uint32_t pin) noexcept {
+    if (!this->is_initialized) {
+        return false;
+    }
+    if (pin >= PINS_COUNT) {
+        return false;
+    }
+    if (!this->EnsurePwmPeripheralOwnership()) {
+        return false;
+    }
+    this->StopPlayback(false);
+
+    this->arduino_pin = pin;
+    this->nrf_pin = g_ADigitalPinMap[pin];
+
+    uint32_t pins[NRF_PWM_CHANNEL_COUNT] = {
+        this->nrf_pin,
+        NRF_PWM_PIN_NOT_CONNECTED,
+        NRF_PWM_PIN_NOT_CONNECTED,
+        NRF_PWM_PIN_NOT_CONNECTED
+    };
+
+    nrf_pwm_pins_set(_PWMInstance, pins); // must set pins before enabling
+    nrf_pwm_enable(_PWMInstance);
+    nrf_pwm_configure(_PWMInstance, TONE_PWM_CONFIG::clock_frequency, TONE_PWM_CONFIG::pwm_mode, this->time_period);
+    nrf_pwm_decoder_set(_PWMInstance, TONE_PWM_CONFIG::load_mode, TONE_PWM_CONFIG::step_mode);
+    nrf_pwm_shorts_set(_PWMInstance, this->shorts);
+    nrf_pwm_int_set(_PWMInstance, 0);
+
+    // static sequence value ... This is the value actually used
+    // during playback ... do NOT modify without semaphore *and*
+    // having ensured playback has stopped!
+    static uint16_t seq_values; // static value 
+    seq_values = this->duty_with_polarity;
+
+    nrf_pwm_seq_ptr_set(_PWMInstance, 0, &seq_values);
+    nrf_pwm_seq_ptr_set(_PWMInstance, 1, &seq_values);
+    nrf_pwm_seq_cnt_set(_PWMInstance, 0, 1);
+    nrf_pwm_seq_cnt_set(_PWMInstance, 1, 1);
+
+    nrf_pwm_seq_refresh_set(_PWMInstance, 0, seq0_refresh);
+    nrf_pwm_seq_refresh_set(_PWMInstance, 1, seq1_refresh);
+    nrf_pwm_seq_end_delay_set(_PWMInstance, 0, 0);
+    nrf_pwm_seq_end_delay_set(_PWMInstance, 1, 0);
+    nrf_pwm_loop_set(_PWMInstance, loop_count);
+
+    nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_STOPPED);
+    nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_SEQSTARTED0);
+    nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_SEQSTARTED1);
+    nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_SEQEND0);
+    nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_SEQEND1);
+    nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_PWMPERIODEND);
+    nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_LOOPSDONE);
+    return true;
 }
+bool TONE_PWM_CONFIG::StartPlayback(void) noexcept {
+    if (!this->is_initialized) {
```

Comment:
please remove this, this class is internal, we should only guard again API with external usage only.

Reply:
I disagree.  An abstraction should prevent common misuse.

But, for you, I did this change anyways.